packages/@lwc/engine-core/src/framework/rendering.ts

```diff
@@ -0,0 +1,660 @@
+/*
+ * Copyright (c) 2018, salesforce.com, inc.
+ * All rights reserved.
+ * SPDX-License-Identifier: MIT
+ * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
+ */
+import {
+    ArrayPush,
+    assert,
+    create,
+    isArray,
+    isNull,
+    isTrue,
+    isUndefined,
+    keys,
+    KEY__SHADOW_RESOLVER,
+} from '@lwc/shared';
+
+import { hasDynamicChildren } from './api';
+import { getComponentInternalDef } from './def';
+import { markComponentAsDirty } from './component';
+import { getUpgradableConstructor } from './upgradable-element';
+import { patchElementWithRestrictions, unlockDomMutation, lockDomMutation } from './restrictions';
+import { EmptyArray } from './utils';
+import {
+    createVM,
+    getAssociatedVMIfPresent,
+    VM,
+    ShadowMode,
+    RenderMode,
+    rerenderVM,
+    VMState,
+    runConnectedCallback,
+    appendVM,
+    removeVM,
+    getRenderRoot,
+    LWCDOMMode,
+} from './vm';
+import {
+    VNode,
+    VCustomElement,
+    VElement,
+    VNodes,
+    Key,
+    VText,
+    VComment,
+    VBaseElement,
+    VNodeType,
+} from './vnode';
+
+import { applyEventListeners } from './modules/events';
+import { patchAttributes } from './modules/attrs';
+import { patchProps } from './modules/props';
+import { patchClassAttribute } from './modules/computed-class-attr';
+import { patchStyleAttribute } from './modules/computed-style-attr';
+import { applyStaticClassAttribute } from './modules/static-class-attr';
+import { applyStaticStyleAttribute } from './modules/static-style-attr';
+
+interface KeyToIndexMap {
+    [key: string]: number;
+}
+
+const SVG_NAMESPACE = 'http://www.w3.org/2000/svg';
+
+export function patchChildren(parentElm: ParentNode, oldCh: VNodes, newCh: VNodes): void {
+    if (hasDynamicChildren(newCh)) {
+        updateDynamicChildren(parentElm, oldCh, newCh);
+    } else {
+        updateStaticChildren(parentElm, oldCh, newCh);
+    }
+}
+
+function patch(n1: VNode | null, n2: VNode, parent: ParentNode, anchor: Node | null) {
```

Comment:
@pmdartus something feels off about this function. basically is doing 2 different things based on the parameters. if `n1` is null, then is mounting (create+insert right?), otherwise patching (I'm still trying to figure out the fix me comment). Then in the `process*` we make the same differentiation with a condition for `isNull` (mount/patch).

What do you think if we separate the `patch` logic into `patch` and `mount`, and also the `process*` into `patch*` and `mount*` (you already have them for Element and CustomeElements)? (imho) Such a way is clearer, and we can skip the multiple `isNull` check which are on the critical path of the diffing algo.